File: agencia/src/services/vector_clock.py
import threading
from typing import List


class RelogioVetorial:
    def __init__(self, id_agencia: int, numero_agencias: int):
        self.id_agencia = id_agencia
        self.vetor = [0] * numero_agencias
        self._lock = threading.Lock()

    def evento_local(self) -> List[int]:
        with self._lock:
            self.vetor[self.id_agencia] += 1
            return list(self.vetor)

    def ao_enviar(self) -> List[int]:
        with self._lock:
            self.vetor[self.id_agencia] += 1
            return list(self.vetor)

    def ao_receber(self, vetor_recebido: List[int]) -> List[int]:
        with self._lock:
            for i in range(len(self.vetor)):
                self.vetor[i] = max(self.vetor[i], vetor_recebido[i])
            self.vetor[self.id_agencia] += 1
            return list(self.vetor)

File: agencia-express/src/services/vectorClock.js
class RelogioVetorial {
  constructor(idAgencia, numeroAgencias) {
    this.idAgencia = idAgencia;
    this.vetor = new Array(numeroAgencias).fill(0);
  }

  eventoLocal() {
    this.vetor[this.idAgencia] += 1;
    return [...this.vetor];
  }

  aoEnviar() {
    this.vetor[this.idAgencia] += 1;
    return [...this.vetor];
  }

  aoReceber(vetorRecebido) {
    for (let i = 0; i < this.vetor.length; i++) {
      this.vetor[i] = Math.max(this.vetor[i], vetorRecebido[i]);
    }
    this.vetor[this.idAgencia] += 1;
    return [...this.vetor];
  }
}

export default RelogioVetorial;
